pkg/controller/jobs/pod/pod_webhook.go

```diff
@@ -0,0 +1,155 @@
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+
+var _ webhook.CustomDefaulter = &PodWebhook{}
+
+func (w *PodWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	pod := fromObject(obj)
+	log := ctrl.LoggerFrom(ctx).WithName("pod-webhook").WithValues("pod", klog.KObj(pod))
+	log.V(5).Info("Applying defaults")
+
+	if IsPodOwnerManagedByQueue(pod) {
+		log.V(5).Info("Pod owner is managed by queue, skipping")
+		return nil
+	}
+
+	// Check for pod label selector match
+	podSelector, err := metav1.LabelSelectorAsSelector(w.podSelector)
+	if err != nil {
+		return fmt.Errorf("failed to parse pod selector: %w", err)
+	}
+	if !podSelector.Matches(labels.Set(pod.GetLabels())) {
+		return nil
+	}
+
+	// Get pod namespace and check for namespace label selector match
+	ns := corev1.Namespace{}
+	err = w.client.Get(ctx, client.ObjectKey{Name: pod.GetNamespace()}, &ns)
+	if err != nil {
+		return fmt.Errorf("failed to run mutating webhook on pod %s, error while getting namespace: %w",
+			pod.GetName(),
+			err,
+		)
+	}
+	log.V(5).Info("Found pod namespace", "Namespace.Name", ns.GetName())
+	nsSelector, err := metav1.LabelSelectorAsSelector(w.namespaceSelector)
+	if err != nil {
+		return fmt.Errorf("failed to parse namespace selector: %w", err)
+	}
+	if !nsSelector.Matches(labels.Set(ns.GetLabels())) {
+		return nil
+	}
+
+	if jobframework.QueueName(pod) != "" || w.manageJobsWithoutQueueName {
+		controllerutil.AddFinalizer(obj.(*corev1.Pod), PodFinalizer)
+
+		if pod.Labels == nil {
+			pod.Labels = make(map[string]string)
+		}
+		pod.Labels[ManagedLabelKey] = ManagedLabelValue
+
+		if pod.gateIndex() == gateNotFound {
+			log.V(5).Info("Adding gate")
+			pod.Spec.SchedulingGates = append(pod.Spec.SchedulingGates, corev1.PodSchedulingGate{Name: SchedulingGateName})
+		}
+	}
+
+	return nil
+}
+
+// +kubebuilder:webhook:path=/validate--v1-pod,mutating=false,failurePolicy=ignore,sideEffects=None,groups="",resources=pods,verbs=create;update,versions=v1,name=vpod.kb.io,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &PodWebhook{}
+
+func (w *PodWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	pod := fromObject(obj)
+	log := ctrl.LoggerFrom(ctx).WithName("pod-webhook").WithValues("pod", klog.KObj(pod))
+	log.V(5).Info("Validating create")
+	allErrs := jobframework.ValidateCreateForQueueName(pod)
+	return nil, allErrs.ToAggregate()
+}
+
+func (w *PodWebhook) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	oldPod := fromObject(oldObj)
+	newPod := fromObject(newObj)
+	log := ctrl.LoggerFrom(ctx).WithName("pod-webhook").WithValues("pod", klog.KObj(newPod))
+	log.V(5).Info("Validating update")
+	allErrs := jobframework.ValidateUpdateForQueueName(oldPod, newPod)
```

Comment:
Validating the pod whether schedulingGate isn't removed even though the pod isn't admitted. Because in pod integration, the reconciler will remove the no-admitted pod.

Reply:
How do we know if the pod is admitted? we should avoid looking at other objects from a Webhook, as it could lead to race conditions.

Reply:
Uhm. It sounds reasonable. 
Once pods are removed, we would expect that external-controllers or dispatchers will re-create pods.

Reply:
If the schedulingGate is removed before kueue admits the workload, the reconciler would stop (terminate) the pod.
I think there is already a unit test for that.

Reply:
> I think there is already a unit test for that.

@alculquicondor I can not see the case in the `TestReconciler`. Does that mean another place?

Reply:
@achernevskii can you add a test case for a workload that gets the scheduling gate removed, but the workload is not admitted?

Reply:
Added the following test case: 

https://github.com/kubernetes-sigs/kueue/blob/1c7a932045462bf96865a33213d228549dd02634/pkg/controller/jobs/pod/pod_controller_test.go#L341-L373